skbio/stats/composition.py

```diff
@@ -0,0 +1,384 @@
+    mat = mat / mat.sum(axis=1, keepdims=True)
+    return mat.squeeze()
+
+
+def multiplicative_replacement(mat, delta=None):
+    """Replace all zeros with small non-zero values.
+
+    It uses the multiplicative replacement strategy, replacing zeros
+    with small positive values and performing a closure operation so
+    that the compositions still add up to 1
+
+    Parameters
+    ----------
+    mat: array_like
+       a matrix of proportions where
+       rows = compositions and
+       columns = components
+
+    Returns
+    -------
+    numpy.ndarray, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skbio.stats.composition import multiplicative_replacement
+    >>> X = np.array([[.2,.4,.4, 0],[0,.5,.5,0]])
+    >>> multiplicative_replacement(X)
+    array([[ 0.1875,  0.375 ,  0.375 ,  0.0625],
+           [ 0.0625,  0.4375,  0.4375,  0.0625]])
+
+    """
+    mat = np.atleast_2d(mat)
+    if mat.ndim > 2:
+        raise ValueError("Input matrix can only have two dimensions or less")
+    if np.any(mat < 0):
+        raise ValueError("Cannot have negative proportions")
+    if not np.all(np.isclose(mat.sum(axis=1), 1)):
+        raise ValueError("Rows need to sum up to 1")
+
+    z_mat = (mat == 0)
+
+    num_samps, num_feats = mat.shape
+    tot = z_mat.sum(axis=1, keepdims=True)
+
+    if delta is None:
+        delta = (1. / num_feats)**2
+
+    zcnts = 1 - tot * delta
+    mat = np.where(z_mat, delta, zcnts * mat)
+    return mat.squeeze()
+
+
+def perturb(x, y):
+    r"""
+    Performs the perturbation operation.
+
+    This operation is defined as
+    :math:`x \oplus y = C[x_1 y_1, ..., x_D y_D]`
+
+    :math:`C[x]` is the closure operation defined as
+    :math:`C[x] = [\frac{x_1}{\sum x},...,\frac{x_D}{\sum x}]`
+    for some :math:`D` dimensional real vector :math:`x` and
+    :math:`D` is the number of components for every composition.
+
+    Parameters
+    ----------
+    x : array_like, float
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+    y : array_like, float
+        a matrix of proportions where
+        rows = compositions and
+        columns = components
+
+    Returns
+    -------
+    numpy.ndarray, np.float64
+       A matrix of proportions where all of the values
+       are nonzero and each composition (row) adds up to 1
+
+    Notes
+    -----
+    - All of the values in x and y must be non negative
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skbio.stats.composition import perturb
+    >>> x = np.array([.1,.3,.4, .2])
+    >>> y = np.array([1./6,1./6,1./3,1./3])
+    >>> perturb(x,y)
+    array([ 0.0625,  0.1875,  0.5   ,  0.25  ])
+
+    """
+    x, y = closure(x), closure(y)
+    return closure(x * y)
+
+
+def perturb_inv(x, y):
```

Comment:
I am a little on the edge about that, here's why.

If you are stacking many of these operations together, adding in additional parameters to specify inverses can be messy really quickly.  Its a similar reason why there is a `np.multiply` and a `np.divide` function.  If there was only a single `np.multiply` function with a parameter specifying division, you can imagine that matrix manipulations can become very messy.

I contemplated having a `Composition` object and overloading the `+`, `-`, `*` and `\` operators.
But I'm still in favor of having this scheme, because the operator overloading can get confusing and the data structures are a bit more transparent.